### PR TITLE
Moby container runtime

### PR DESCRIPTION
**Pull request by jackfrancis**
**What this PR does / why we need it**: 

Deprecates docker-engine in favor of moby.

See:

https://docs.microsoft.com/en-us/azure/iot-edge/how-to-install-iot-edge-linux#register-microsoft-key-and-software-repository-feed

Includes a prototype of a feature-flag-enabled VHD CI process. This prototype assumes that at CI time, we include an environment variable string that includes a list of enabled feature flags. E.g.:

`FEATURE_FLAGS="feature-foo,feature-bar"`

And then VHD CI would evaluate like this:

`if [[ $FEATURE_FLAGS == *"feature-foo"* ]]; then`...

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**If applicable**:
- [ ] documentation
- [ ] unit tests
- [ ] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by jackfrancis** [comment]
Testing is currently blocked on:

```
$ kubectl describe hpa php-apache-kubernetes-eastus-50779-53479
Name:                                                  php-apache-kubernetes-eastus-50779-53479
Namespace:                                             default
Labels:                                                <none>
Annotations:                                           <none>
CreationTimestamp:                                     Mon, 24 Sep 2018 17:28:06 +0000
Reference:                                             Deployment/php-apache-kubernetes-eastus-50779-53479
Metrics:                                               ( current / target )
  resource cpu on pods  (as a percentage of request):  <unknown> / 5%
Min replicas:                                          1
Max replicas:                                          10
Conditions:
  Type           Status  Reason                   Message
  ----           ------  ------                   -------
  AbleToScale    True    SucceededGetScale        the HPA controller was able to get the target's current scale
  ScalingActive  False   FailedGetResourceMetric  the HPA was unable to compute the replica count: unable to get metrics for resource cpu: no metrics returned from resource metrics API
Events:
  Type     Reason                        Age              From                       Message
  ----     ------                        ----             ----                       -------
  Warning  FailedGetResourceMetric       2s (x3 over 1m)  horizontal-pod-autoscaler  unable to get metrics for resource cpu: no metrics returned from resource metrics API
  Warning  FailedComputeMetricsReplicas  2s (x3 over 1m)  horizontal-pod-autoscaler  failed to get cpu utilization: unable to get metrics for resource cpu: no metrics returned from resource metrics API
```

**Comment by jackfrancis** [comment]
FYI same symptom above w/ Azure CNI or Kubenet.

**Comment by khenidak** [comment]
@jackfrancis .. @cpuguy83 graciously agreed to help with this one. I shared with him the high level details. 

@ritazh FYI

**Comment by cpuguy83** [comment]
Ok... here's the problem.


```
Registration of the Docker container factory failed: failed to validate Docker info: version string "dev" doesn't match expected regular expression: "(\d+)\.(\d+)\.(\d+)"
```

This error is from the kubelet when it is registering cadvisor metrics handlers (FYI, kubelet uses an internal cadvisor to collect metrics for the "docker" runtime).

When building we need to set a version to use otherwise the version output is:

```
Server:
 Engine:
  Version:      dev
  API version:  1.38 (minimum version 1.12)
  Go version:   go1.10.3
  Git commit:   09f5e9d
  Built:        Wed Sep 19 16:55:45 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

**Comment by cpuguy83** [comment]
To set the version, when building moby you must specify `-ldflags="-X dockerversion.Version=<version>`.
This will instruct the compiler to overwrite the default variable set in `github.com/moby/moby/dockerversion.Version`

**Comment by cpuguy83** [comment]
Or just hardcode it...

**Comment by khenidak** [comment]
Thanks I will revisit the build process and make sure version is attached

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by khenidak** [comment]
> To set the version, when building moby you must specify `-ldflags="-X dockerversion.Version=<version>`.
> This will instruct the compiler to overwrite the default variable set in `github.com/moby/moby/dockerversion.Version`

for future us it is actually `make VERSION=<version>` the variable above is set by scripts.

**Comment by jackfrancis** [comment]
Initial tests of Windows hybrid clusters against this changeset are 💯

File: parts/k8s/kubernetesmastercustomdata.yml [hunk 1]
```diff
@@ -116,8 +116,9 @@ write_files:
     {{if .MasterProfile.IsCoreOS}}
     ExecStart=/usr/bin/env PATH=${TORCX_BINDIR}:${PATH} ${TORCX_BINDIR}/dockerd --host=fd:// --containerd=/var/run/docker/libcontainerd/docker-containerd.sock --storage-driver=overlay2 --bip={{WrapAsParameter "dockerBridgeCidr"}} $DOCKER_SELINUX $DOCKER_OPTS $DOCKER_CGROUPS $DOCKER_OPT_BIP $DOCKER_OPT_MTU $DOCKER_OPT_IPMASQ
     {{else}}
-    ExecStart=/usr/bin/docker daemon -H fd:// --storage-driver=overlay2 --bip={{WrapAsParameter "dockerBridgeCidr"}}
+    ExecStart=/usr/bin/dockerd -H fd:// --storage-driver=overlay2 --bip={{WrapAsParameter "dockerBridgeCidr"}}
```
Comment (jackfrancis) [inline parts/k8s/kubernetesmastercustomdata.yml]: This change is backwards-compatible w/ `docker-engine` v1.13.1

File: parts/k8s/kubernetesagentcustomdata.yml [hunk 2]
```diff
@@ -100,13 +100,14 @@ write_files:
...
     {{end}}
+
 - path: "/etc/systemd/system/docker.service.d/exec_start.conf"
   permissions: "0644"
   owner: "root"
   content: |
     [Service]
     ExecStart=
-    ExecStart=/usr/bin/docker daemon -H fd:// --storage-driver=overlay2 --bip={{WrapAsParameter "dockerBridgeCidr"}}
+    ExecStart=/usr/bin/dockerd -H fd:// --storage-driver=overlay2 --bip={{WrapAsParameter "dockerBridgeCidr"}}
```
Comment (jackfrancis) [inline parts/k8s/kubernetesagentcustomdata.yml]: This change is backwards-compatible w/ docker-engine v1.13.1

File: pkg/api/vlabs/types.go [hunk 3]
```diff
@@ -277,7 +277,7 @@ type KubernetesConfig struct {
 	UserAssignedID                  string            `json:"userAssignedID,omitempty"`
 	UserAssignedClientID            string            `json:"userAssignedClientID,omitempty"` //Note: cannot be provided in config. Used *only* for transferring this to azure.json.
 	CustomHyperkubeImage            string            `json:"customHyperkubeImage,omitempty"`
-	DockerEngineVersion             string            `json:"dockerEngineVersion,omitempty"`
+	DockerEngineVersion             string            `json:"dockerEngineVersion,omitempty"` // Deprecated
```
Comment (CecileRobertMichon) [inline pkg/api/vlabs/types.go]: we might want to add some sort of validation/warning message that DockerEngineVersion is deprecated and Docker is replaced by Moby for customers still passing in a Docker engine version. Thinking of customers with working apimodels specifying a Docker Engine Version that always use the same apimodels.

File: parts/k8s/kubernetesprovisionsource.sh [hunk 4]
```diff
@@ -18,6 +18,9 @@ ERR_DOCKER_DOWNLOAD_TIMEOUT=21 # Timout waiting for docker download(s)
 ERR_DOCKER_KEY_DOWNLOAD_TIMEOUT=22 # Timeout waiting to download docker repo key
 ERR_DOCKER_APT_KEY_TIMEOUT=23 # Timeout waiting for docker apt-key
 ERR_DOCKER_START_FAIL=24 # Docker could not be started by systemctl
+ERR_MOBY_APT_LIST_TIMEOUT=25 # Timeout waiting for moby apt sources
+ERR_MS_GPG_KEY_DOWNLOAD_TIMEOUT=26 # Timeout waiting for MS GPG key download
+ERR_MOBY_INSTALL_TIMEOUT=27 # Timeout waiting for moby install
```
[the remark this answers is not in the document]
Comment (jackfrancis) [reply parts/k8s/kubernetesprovisionsource.sh]: done

File: pkg/api/vlabs/types.go (hunk 3 above)
Comment (jackfrancis) [reply pkg/api/vlabs/types.go]: good idea, done

File: pkg/api/vlabs/validate.go [hunk 5]
```diff
@@ -281,6 +281,10 @@ func (a *Properties) validateOrchestratorProfile(isUpdate bool) error {
 						return errors.Errorf("standard loadBalancerSku should exclude master nodes. Please set KubernetesConfig \"ExcludeMasterFromStandardLB\" to \"true\"")
 					}
 				}
+
+				if o.KubernetesConfig.DockerEngineVersion != "" {
+					log.Warnf("docker-engine is deprecated, but you passed in a dockerEngineVersion configuration. This will be ignored.")
```
Comment (CecileRobertMichon) [inline pkg/api/vlabs/validate.go]: let's mention Moby in here since it's nowhere in docs

File: parts/k8s/addons/kubernetesmasteraddons-azure-npm-daemonset.yaml [hunk 6]
```diff
@@ -75,7 +75,7 @@ spec:
         beta.kubernetes.io/os: linux
       containers:
         - name: azure-npm
-          image: containernetworking/azure-npm:v1.0.9
+          image: containernetworking/azure-npm:v1.0.13
```
Comment (CecileRobertMichon) [inline parts/k8s/addons/kubernetesmasteraddons-azure-npm-daemonset.yaml]: why is this version changing and how is this related to moby?

File: parts/k8s/addons/kubernetesmasteraddons-blobfuse-flexvolume-installer.yaml [hunk 7]
```diff
@@ -25,7 +25,7 @@ spec:
     spec:
       containers:
       - name: blobfuse-flexvol-installer
-        image: andyzhangx/blobfuse-flexvol-installer:1.0.6
```
Comment (CecileRobertMichon) [inline parts/k8s/addons/kubernetesmasteraddons-blobfuse-flexvolume-installer.yaml]: ditto

File: parts/k8s/addons/kubernetesmasteraddons-smb-flexvolume-installer.yaml [hunk 8]
```diff
@@ -25,7 +25,7 @@ spec:
     spec:
       containers:
       - name: smb-flexvol-installer
-        image: andyzhangx/smb-flexvol-installer:1.0.2
```
Comment (CecileRobertMichon) [inline parts/k8s/addons/kubernetesmasteraddons-smb-flexvolume-installer.yaml]: and here

File: parts/k8s/kubernetesagentresourcesvmas.t [hunk 9]
```diff
@@ -174,7 +174,7 @@
       "identity": {
         "type": "userAssigned",
         "userAssignedIdentities": {
-          "[resourceId('Microsoft.ManagedIdentity/userAssignedIdentities/', variables('userAssignedID'))]":{}
+          "[variables('userAssignedIDReference')]":{}
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetesagentresourcesvmas.t]: this might need to be rebased

File: parts/k8s/addons/kubernetesmasteraddons-azure-npm-daemonset.yaml (hunk 6 above)
Comment (jackfrancis) [reply parts/k8s/addons/kubernetesmasteraddons-azure-npm-daemonset.yaml]: hmm, rebase errata

**Comment by cpuguy83** [comment]
🎉
